r/tools/check-versions.R

```diff
@@ -20,6 +20,16 @@ args <- commandArgs(TRUE)
 # TESTING is set in test-check-version.R; it won't be set when called from configure
 test_mode <- exists("TESTING")
 
+release_version_supported <- function(r_version, cpp_version) {
+  r_version <- package_version(r_version)
+  cpp_version <- package_version(cpp_version)
+  major <- function(x) as.numeric(x[1, 1])
+
+  # Currently the last three released versions of Arrow C++
+  # (>= 13.0.0 at the time this check was added)
+  major(cpp_version) %in% (major(r_version) - 0:2)
```

Comment:
This feels brittle and opaque. IMO it would be clearer if we had an explicit version that the C++ library must be greater than or equal to. Then it's just `cpp_version >=` that.

Reply:
Right, but right now, there only one preprocessor statement, and it's tied to a specific version (15). So after 16.0 comes out, there's no reason to "drop" support for 13.0, it still works. Even when there are more, and they add maintenance burden, you drop support when you remove the preprocessor statements, which are tied to versions, so you're still talking about a minimum version you support. If you decide to be aggressive about dropping them, that's fine, I just think it's cleaner to express it as "libarrow >= X.Y" than "last three versions". Easier for users to understand, easier to make sure you're updating all the right places, etc.

Reply:
I agree that this line is quite opaque. It took me a decent bit of parsing to figure out what it's actually doing here.